Enable dynamic input switching

utzcoz opened the pull request:
The `frameworks_base` part to enable dynamic input switching. For the whose view, please visit the issue [Enable dynamic input switching](https://github.com/maruos/maruos/issues/97).

utzcoz committed:
Enable dynamic input switching

Drop all input events except touch screen/pad/nativation events
when sys.maruos.desktop.interactive is true, which will be set to true
after starting lxc container succeed with public presentation, and set
to false after disconnecting public presentation or stopping lxc.

Signed-off-by: utzcoz <utzcoz@outlook.com>

pdsouza commented on services/core/java/com/android/server/mperspective/PerspectiveService.java:
```diff
@@ -364,24 +391,25 @@ public void handleMessage(Message msg) {
                     updateDesktopState();
                     break;
+                case MSG_DISABLE_DESKTOP_INTERACTIVE:
+                    updateDesktopInteractiveStateLocked(false);
```
We need to run this in a `synchronized (mLock)` block to prevent concurrency races. You can create a method in this handler class like `updateDesktopInteractiveStateInternal()` like we do for `startDesktopPerspectiveInternal()` and use a synchronized block there.

pdsouza commented on services/core/java/com/android/server/mperspective/PerspectiveService.java:
```diff
@@ -364,24 +391,25 @@ public void handleMessage(Message msg) {
+                    mPublicPresentationDisplayId = displayId;
+                    if (mPublicPresentationAutoStart) {
+                        Log.i(TAG, "Public presentation display added, scheduling desktop start...");
                         scheduleStartDesktopPerspective();
```
I think we will need `scheduleEnableDesktopInteractive()` here similar to the logic in `onDisplayRemoved()` to handle the case where the desktop is already in `STATE_RUNNING` and needs to have the input enabled when a public presentation display is added.

pdsouza reviewed (changes requested):
Overall looks good to me. But I think we must fix the scenario where the desktop is already in `STATE_RUNNING` and a public presentation display is added after previously being disconnected.

pdsouza commented:
Hi @utzcoz, I think I found one more case we need to handle.

Consider this scenario:
1. Connect to public presentation display. The desktop will auto-start in interactive mode (input enabled for Linux, disabled for Android).
2. Disconnect the public presentation display. The desktop will go to background mode (input disabled for Linux, enabled for Android).
3. *Re-connect the public presentation display*. But from this code it looks like desktop will still stay in background mode? Since the container is already in `STATE_RUNNING` we will not run the `enabIeInput` logic in `updateDesktopStateLocked` ([ref](https://github.com/maruos/android_platform_frameworks_base/pull/5/files#diff-85fab8f12f3a3954e15f12dedcd4b711R171)). I think we need to add code in `onDisplayAdded` to make sure we enable input for the desktop.

utzcoz commented:
@pdsouza You're right, I have fixed the problem, and squashed all commits to a main commit for cleaning up.

pdsouza reviewed (approved):
Thanks for fixing all issues and squashing your commits. This looks good!

pdsouza commented on core/java/android/view/ViewRootImpl.java:
```diff
@@ -4204,6 +4205,19 @@ protected void onDeliverToNext(QueuedInputEvent q) {
+            // In this occasion, we will use (bluetooth or usb) keyboard and mouse
+            // in Linux, and use touch screen/pad/navigation in Android for controlling
+            // the state of Linux manually when using Android cast to show Linux.
+            boolean isTouchEvent = q.mEvent != null
```
From my testing it looks like the Android software "back" button is being ignored when the desktop input is enabled. Maybe it is falling into this check?

utzcoz replied:
Yes, the software `back` button is a physical event, so I change the logic to drop external input events and keep all built-in input events which will fix this problem and simplify the code.

pdsouza commented on services/core/java/com/android/server/mperspective/PerspectiveService.java:
```diff
@@ -332,6 +351,10 @@ private void startDesktopPerspectiveInternal() {
+                // After starting desktop, we should make sure the desktop
+                // input is disabled. If in some occasion, we want to use
+                // input in desktop, please enable it manually.
+                updateDesktopInteractiveStateInternal(false);
```
I think we need to check whether a public presentation display is available and then set the interactive state based on that. If the user stops the desktop manually and then starts it again manually while a public presentation display is connected, this will disable the input for the desktop even though it should be enabled since a display is available. The user will need to disconnect and re-connect the display to fix this.

utzcoz replied:
I remove logic to `updateDesktopStateLocked`, and combine public presentation display connecting state and desktop running state to set desktop interactive state.

pdsouza reviewed (commented):
I found two small issues while testing. Once these are fixed we should be in good shape!

utzcoz committed:
Force showing IME when desktop is interactive

Defaulty, the Android don't show IME when hardware
keyboard is connected. But when desktop is interactive,
the hardware keyboard event is directed to desktop,
so we need showing IME when desktop is interactive
for better user experience.

Signed-off-by: utzcoz <utzcoz@outlook.com>

utzcoz commented:
@pdsouza I add a new patch to force showing IME when desktop is interactive.

pdsouza commented:
@utzcoz Merged - thanks for your patience in getting all these PRs in!
